mcs/class/corlib/System/MulticastDelegate.cs

```diff
@@ -112,167 +109,141 @@ public sealed override int GetHashCode ()
 		// </summary>
 		public sealed override Delegate[] GetInvocationList ()
 		{
-			MulticastDelegate d;
-			d = (MulticastDelegate) this.Clone ();
-			for (d.kpm_next = null; d.prev != null; d = d.prev)
-				d.prev.kpm_next = d;
-
-			if (d.kpm_next == null) {
-				MulticastDelegate other = (MulticastDelegate) d.Clone ();
-				other.prev = null;
-				other.kpm_next = null;				
-				return new Delegate [1] { other };
-			}
-
-			var list = new List<Delegate> ();
-			for (; d != null; d = d.kpm_next) {
-				MulticastDelegate other = (MulticastDelegate) d.Clone ();
-				other.prev = null;
-				other.kpm_next = null;
-				list.Add (other);
-			}
-
-			return list.ToArray ();
+			return delegates;
 		}
 
 		// <summary>
-		//   Combines this MulticastDelegate with the (Multicast)Delegate `follow'.
+		//   Combines this MulticastDelegate with the (Multicast)Delegate `other'.
 		//   This does _not_ combine with Delegates. ECMA states the whole delegate
 		//   thing should have better been a simple System.Delegate class.
 		//   Compiler generated delegates are always MulticastDelegates.
 		// </summary>
-		protected sealed override Delegate CombineImpl (Delegate follow)
+		protected override sealed Delegate CombineImpl (Delegate v)
 		{
-			if (follow == null)
-				return this;
 
-			MulticastDelegate combined, orig, clone;
+			if (v == null)
+				return this;
 
-			if (this.GetType() != follow.GetType ())
-				throw new ArgumentException (Locale.GetText ("Incompatible Delegate Types. First is {0} second is {1}.", this.GetType ().FullName, follow.GetType ().FullName));
+			MulticastDelegate other = (MulticastDelegate) v;
 
-			combined = (MulticastDelegate)follow.Clone ();
-			combined.SetMulticastInvoke ();
+			MulticastDelegate ret = InternalAllocLike_internal (this);
 
-			for (clone = combined, orig = ((MulticastDelegate)follow).prev; orig != null; orig = orig.prev) {
-				
-				clone.prev = (MulticastDelegate)orig.Clone ();
-				clone = clone.prev;
-			}
+			if (delegates == null && other.delegates == null) {
+				ret.delegates = new Delegate [2] { this, other };
+			} else if (delegates == null) {
+				ret.delegates = new Delegate [1 + other.delegates.Length];
 
-			clone.SetMulticastInvoke ();
-			clone.prev = (MulticastDelegate)this.Clone ();
+				ret.delegates [0] = this;
+				Array.Copy (other.delegates, 0, ret.delegates, 1, other.delegates.Length);
+			} else if (other.delegates == null) {
+				ret.delegates = new Delegate [delegates.Length + 1];
```

Comment:
Array.Resize copies the array and adjust the size up or down.
It doesn't perform an inplace change.
The API name is stupid.